Support mTLS connection-level capabilities and RPC access filtering in C++ [run-systemtest]

vekterli opened the pull request:
@havardpe @bjorncs please review

Adds the following:
 * Named capabilities and capability sets that represent (respectively)
   a single Vespa access API (such as Document API, search API etc)
   or a concrete subset of individual capabilities that make up a
   particular Vespa service (such as a content node).
 * A new `capabilities` array field to the mTLS authorization policies
   that allows for constraining what requests sent over a particular
   connection are allowed to actually do. Capabilities are referenced
   by name and may include any combination of capability sets and
   individual capabilities. If multiple capabilities/sets are configured,
   the resulting set of capabilities is the union set of all of them.
 * An FRT RPC-level access filter that can be set up as part of RPC
   method definitions. If set, filters are invoked prior to RPC methods.
 * A new `PERMISSION_DENIED` error code to FRT RPC that is invoked if
   an access filter denies a request.

This also GCs the unused `AssumedRoles` concept which is now deprecated
in favor of capabilities.

This PR does not wire access filtering to any existing RPC method definitions;
this will be added when the base functionality is in.

Note: this is **not yet** a public or stable API, and capability
names/semantics may change at any time!

vekterli committed:
Support mTLS connection-level capabilities and RPC access filtering in C++

Adds the following:
 * Named capabilities and capability sets that represent (respectively)
   a single Vespa access API (such as Document API, search API etc)
   or a concrete subset of individual capabilities that make up a
   particular Vespa service (such as a content node).
 * A new `capabilities` array field to the mTLS authorization policies
   that allows for constraining what requests sent over a particular
   connection are allowed to actually do. Capabilities are referenced
   by name and may include any combination of capability sets and
   individual capabilities. If multiple capabilities/sets are configured,
   the resulting set of capabilities is the union set of all of them.
 * An FRT RPC-level access filter that can be set up as part of RPC
   method definitions. If set, filters are invoked prior to RPC methods.
 * A new `PERMISSION_DENIED` error code to FRT RPC that is invoked if
   an access filter denies a request.

This also GCs the unused `AssumedRoles` concept which is now deprecated
in favor of capabilities.

Note: this is **not yet** a public or stable API, and capability
names/semantics may change at any time.

vekterli committed:
Clarify why we're setting the return handler prior to filter invocation

havardpe commented on fnet/src/vespa/fnet/frt/require_capability.h:
```diff
@@ -0,0 +1,21 @@
+ * context that contains, at minimum, a given set of capabilities. If one or more
+ * required capabilities are missing, the request is denied.
+ */
+class FRT_RequireCapability final : public FRT_RequestAccessFilter {
```
consider plural in class name

vekterli replied:
Fixed

havardpe commented on vespalib/src/tests/net/tls/capabilities/capabilities_test.cpp:
```diff
@@ -0,0 +1,152 @@
+    EXPECT_EQUAL(Capability::content_storage_api().to_string(), "Capability(vespa.content.storage_api)");
+}
+
+TEST("All known capabilities can be looked up by name") {
```
consider also matching the values against the actual capabilities

vekterli replied:
Added checking of all returned values

havardpe reviewed (approved):
some thoughts:
the CapabilityId enum might not need to be part of the API
bit_id and as_bit_set could be named value/as_set to 'hide' the implementation

vekterli committed:
Refactor Capability(Set) and add more testing

Hide all nitty-gritty details of how capabilities map to internal
bit set positions by making more of Capability private and only
allowing CapabilitySet to see how the sausages are made. Move all
bit set functionality to CapabilitySet, where it really belongs.

vekterli commented:
@havardpe PTAL. Have refactored and moved `Capability` ID functionality around so that it is no longer exposed as part of the API. `bit_id`/`as_bit_set` is no longer available.
